e2store/src/era.rs

```diff
@@ -109,13 +109,19 @@ impl Era {
     }
 
     fn get_block_slot_indexes(slot_index_block_entry: &SlotIndexBlockEntry) -> Vec<u64> {
+        if slot_index_block_entry.slot_index.indices.is_empty() {
+            return vec![];
+        }
+        // minus 8 because the first 8 bytes are the starting slot
+        let beginning_of_index_record = slot_index_block_entry.slot_index.indices[0] - 8;
```

Comment:
this logic doesn't seem right for me. Maybe I'm missing something, so if you could clarify:
1. with current logic, first slot will always be occupied (because check at line 124 will always be false). I don't think that's correct.
2. are we sure that `indices[0]` can't be lower than 8 (in which case we would panic)?

Reply:
> with current logic, first slot will always be occupied (because check at line 124 will always be false). I don't think that's correct.

The first index isn't an offset of 0, it is an offset of 8 because `starting_slot`

> are we sure that indices[0] can't be lower than 8 (in which case we would panic)?

Yes because it will always be 8 bytes higher than `starting_slot`

Reply:
I looked more into this, and after debugging bytes and looking for multiple example, I think I figured it out.

---

First, the indices are negative integers and we should update `indices` field in both `SlotIndexBlock` and `SlotIndexState` to `[i64; X]`. It even says so in the [spec](https://github.com/status-im/nimbus-eth2/blob/613f4a9a50c9c4bd8568844eaffb3ac15d067e56/docs/e2store.md#era-files):

> A SlotIndex record may appear in a stand-alone file which by convention ends with .e2i - in this case, the offset is counted as if the index was appened to its corresponding data file - offsets are thus negative and counted from the end of the data file. In particular, if the index is simply appended to the data file, it does not change in contents.

---

Second, if index (offset) point to the beginning of file (meaning `index == position_of_slot_index_block`), then that block is skipped.
In most cases, first block is present, and will be 8 bytes from the start of file (VersionEntry takes first 8 bytes), resulting in your math working but with the wrong reasoning (it's not because of the `starting_slot` but because of the `VersionEntry`).
However, if first block is skipped (if I'm not mistaken, epoch 819 has starting slot [6701056](https://beaconscan.com/slot/6545408) which is empty), this will not work, because first block already points to the beginning of file and we don't need to subtract 8 bytes.

---

Now, the solution. For this function to work, we need to know at what byte the `SlotIndexBlock` start in the original file, and we have 2 ways of doing so:
1. sum encoded sizes of all entries up to this point (version + blocks + state)
2. know length of the file and subtract encoded sizes of SlotIndexBlock and SlotIndexState entries (which are fixed: `8192*8 + 3*8 = 65560` and `4*8=32`)

---

I think it's a bit hard to explain this in words, but I tried my best. If something is not clear, we can have a call and discuss it.

Reply:
I choose the 2nd option